tools/geom.go

```diff
@@ -0,0 +1,450 @@
+	Fields      map[string]interface{} `json:"fields"`
+	Geometry    []uint8                `json:"geometry"`
+}
+
+type xyzPoint struct {
+	x float64
+	y float64
+	z float64
+}
+
+func rightofEquals(line string) string {
+	return strings.TrimSpace(strings.Split(line, "=")[1])
+}
+
+func dataPairsfromTextBlock(sc *bufio.Scanner, nPairs int, colWidth int, valueWidth int) ([][2]float64, error) {
+	var stride int = valueWidth * 2
+	pairs := [][2]float64{}
+out:
+	for sc.Scan() {
+		line := sc.Text()
+		for s := 0; s < colWidth; {
+			if len(line) > s {
+				val1, err := strconv.ParseFloat(strings.TrimSpace(line[s:s+valueWidth]), 64)
+				if err != nil {
+					return pairs, err
+				}
+				val2, err := strconv.ParseFloat(strings.TrimSpace(line[s+valueWidth:s+stride]), 64)
+				if err != nil {
+					return pairs, err
+				}
+				pairs = append(pairs, [2]float64{val1, val2})
+				if int(len(pairs)) == nPairs {
+					break out
+				}
+			} else {
+				break
+			}
+			s += stride
+		}
+	}
+	return pairs, nil
+}
+
+func getDataPairsfromTextBlock(nDataPairsLine string, sc *bufio.Scanner, colWidth int, valueWidth int) ([][2]float64, error) {
+	pairs := [][2]float64{}
+	for sc.Scan() {
+		line := sc.Text()
+		if strings.HasPrefix(line, nDataPairsLine) {
+			nPairs, err := strconv.Atoi(rightofEquals(line))
+			if err != nil {
+				return pairs, err
+			}
+			pairs, err = dataPairsfromTextBlock(sc, nPairs, colWidth, valueWidth)
+			if err != nil {
+				return pairs, err
+			}
+			break
+		}
+	}
+	return pairs, nil
+}
+
+// distance returns the distance along a straight line in euclidean space
+func distance(p0, p1 [2]float64) float64 {
+	result := math.Sqrt(math.Pow((p1[0]-p0[0]), 2) + math.Pow((p1[1]-p0[1]), 2))
+	return result
+}
+
+// pointAtDistance returns a new point along a straight line in euclidean space
+// at a specified distance
+func pointAtDistance(p0, p1 [2]float64, delta float64) [2]float64 {
+	distanceRatio := delta / distance(p0, p1)
+	newX := (1-distanceRatio)*p0[0] + distanceRatio*p1[0]
+	newY := (1-distanceRatio)*p0[1] + distanceRatio*p1[1]
+	return [2]float64{newX, newY}
+}
+
+// interpZ creates a new point a given distance along a line composed
+// of many segments.
+func interpXY(xyPairs [][2]float64, d float64) [2]float64 {
+	// newPoint is an x, y pair
+	var newPoint [2]float64
+	lineSegments := len(xyPairs) - 1
+	lineLength := 0.0
+
+findLineSegment:
+	for i := 0; i < lineSegments; i++ {
+		p0, p1 := xyPairs[i], xyPairs[i+1]
+		lineLength += distance(p0, p1)
+
+		switch {
+		case lineLength > d:
+			delta := distance(p0, p1) - (lineLength - d)
+			newPoint = pointAtDistance(p0, p1, delta)
+			break findLineSegment
+
+		default:
+			continue
+		}
+
+	}
+
+	return newPoint
+}
+
+// attributeZ using station from cross-section line and gis coordinates
+// Minor bug here caused (I think) where interpolating
+// a station beyond the length of the cutline occurs resulting
+// in a point (0,0). Need to add improved error handling for this.
```

Comment:
I think this would happen if your if statement never comes true. You could add a check after to for loop for that. I also like how you went with a generic `[2]float64` dtype like we discussed earlier!

Reply:
I added an if statement to the interpXY function, which checks if the station of the interpolated point is within a given tolerance of the xy line length (0.1), in which case I return the last point of the xy line. This seems to have resolved this issue, however, I added two additional print statements in case the line station is outside the tolerance or if an interpolated point is still 0,0.